markets/storageadapter/dealpublisher.go

```diff
@@ -0,0 +1,294 @@
+		p.cancelWaitForMoreDeals = nil
+	}
+
+	// Make sure the new deal hasn't been cancelled
+	if pdeal.ctx.Err() != nil {
+		return
+	}
+
+	// Add the new deal to the queue
+	p.pending = append(p.pending, pdeal)
+	log.Infof("add deal with piece CID %s to publish deals queue - %d deals in queue (max queue size %d)",
+		pdeal.deal.Proposal.PieceCID, len(p.pending), p.maxDealsPerPublishMsg)
+
+	// If the maximum number of deals per message has been reached,
+	// send a publish message
+	if uint64(len(p.pending)) >= p.maxDealsPerPublishMsg {
+		log.Infof("publish deals queue has reached max size of %d, publishing deals", p.maxDealsPerPublishMsg)
+		p.publishAllDeals()
+		return
+	}
+
+	// Otherwise wait for more deals to arrive or the timeout to be reached
+	p.waitForMoreDeals()
+}
+
+func (p *DealPublisher) waitForMoreDeals() {
+	// Check if we're already waiting for deals
+	if !p.publishPeriodStart.IsZero() {
+		elapsed := time.Since(p.publishPeriodStart)
+		log.Infof("%s elapsed of / %s until publish deals queue is published",
+			elapsed, p.publishPeriod)
+		return
+	}
+
+	// Set a timeout to wait for more deals to arrive
+	log.Infof("waiting publish deals queue period of %s before publishing", p.publishPeriod)
+	ctx, cancel := context.WithCancel(p.ctx)
+	p.publishPeriodStart = time.Now()
+	p.cancelWaitForMoreDeals = cancel
+
+	go func() {
+		timer := time.NewTimer(p.publishPeriod)
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+		case <-timer.C:
+			p.lk.Lock()
+			defer p.lk.Unlock()
+
+			// The timeout has expired so publish all pending deals
+			log.Infof("publish deals queue period of %s has expired, publishing deals", p.publishPeriod)
+			p.publishAllDeals()
+		}
+	}()
+}
+
+func (p *DealPublisher) publishAllDeals() {
+	// If the timeout hasn't yet been cancelled, cancel it
+	if p.cancelWaitForMoreDeals != nil {
+		p.cancelWaitForMoreDeals()
+		p.cancelWaitForMoreDeals = nil
+		p.publishPeriodStart = time.Time{}
+	}
+
+	// Filter out any deals that have been cancelled
+	p.filterCancelledDeals()
+	deals := p.pending[:]
+	p.pending = nil
+
+	// Send the publish message
+	go p.publishReady(deals)
+}
+
+func (p *DealPublisher) publishReady(ready []*pendingDeal) {
+	if len(ready) == 0 {
+		return
+	}
+
+	deals := make([]market2.ClientDealProposal, 0, len(ready))
+	for _, pd := range ready {
+		deals = append(deals, pd.deal)
+	}
+
+	// Send the publish message
+	msgCid, err := p.publishDealProposals(deals)
+
+	// Signal that each deal has been published
+	for _, pd := range ready {
+		pd := pd
+		go func() {
+			res := publishResult{
+				msgCid: msgCid,
+				err:    err,
+			}
+			select {
+			case <-p.ctx.Done():
+			case pd.Result <- res:
+			}
+		}()
+	}
+}
+
+// Sends the publish message
+func (p *DealPublisher) publishDealProposals(deals []market2.ClientDealProposal) (cid.Cid, error) {
+	log.Infof("publishing %d deals in publish deals queue with piece CIDs: %s", len(deals), pieceCids(deals))
+
+	provider := deals[0].Proposal.Provider
```

Comment:
I'd at least assert that the provider is the same for all deals; there may be a point where we'll want to support multiple miner actors from a single lotus-miner process